MT-25 Created Airport simulation (homework)

AliakseiDziardziyashenka opened the pull request:
Created new project (AirportSimulation) in a new folder named HW9, along with a new branch.

AliakseiDziardziyashenka committed:
MT-25 Created Airport simulation (homework)

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passport.cs:
```diff
@@ -0,0 +1,24 @@
+        }
+    }
+
+    public bool PassportIsValid() => !string.IsNullOrEmpty(PassportNumber) && PassportExpirationDate > DateTime.Today;
```
The naming convention for boolean operations - IsSomethingTrueOfFalse
```suggestion
    public bool IsPasspordValid() => !string.IsNullOrEmpty(PassportNumber) && PassportExpirationDate > DateTime.Today;
```

And the question - do you need a method for this or just a prop? 

```suggestion
   public bool IsValid => !string.IsNullOrEmpty(PassportNumber) && PassportExpirationDate > DateTime.Today;
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Baggage.cs:
```diff
@@ -0,0 +1,6 @@
+
+public class Baggage
+{
+    
```
Do you use it?

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,41 @@
+        string passportNumber = Console.ReadLine();
+
+        Console.Write("Enter Passport Issue Date (yyyy-mm-dd): ");
+        DateTime passportIssueDate = DateTime.Parse(Console.ReadLine());
```
What if the user passes dates that are not in your required format? Please, add validations.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,41 @@
+        DateTime passportIssueDate = DateTime.Parse(Console.ReadLine());
+
+        Console.Write("Enter Passport Expiration Date (yyyy-mm-dd): ");
+        DateTime passportExpirationDate = DateTime.Parse(Console.ReadLine());
```
What if the user passes dates that are not in your required format? Please, add validations.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,41 @@
+        string lastName = Console.ReadLine();
+
+        Console.Write("Enter Date of Birth (yyyy-mm-dd): ");
+        DateTime dateOfBirth = DateTime.Parse(Console.ReadLine());
```
What if the user passes dates that are not in your required format? Please, add validations.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,41 @@
+            Console.WriteLine($"{(int)country} - {country}");
+        }
+
+        int countryChoice = int.Parse(Console.ReadLine());
```
What if the user passes a country not in your required format? Please, add validations and use Enum directly as you decided to use it without initial knowledge (34 line is a bad choice) - Enum also has methods Parse and TryParse - please read the documentation.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passenger.cs:
```diff
@@ -0,0 +1,32 @@
+    private string _firstName;
+    private string _lastName;
+    private DateTime _dateOfBirth;
+    private string _passportNumber;
```
Do you use these private fields?

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passenger.cs:
```diff
@@ -0,0 +1,32 @@
+    
+    public Passenger(string firstName, string lastName, DateTime dateOfBirth, Countries.CountryPassengers country, string passportNumber, Passport passport)
+    {
+        FirstName = firstName;
```
Exactly here, I expected initializing private fields. Your public props have a getter and setter, which you can use directly from every place (assigning in the constructor is not a requirement, but in terms of the private fields and incapsulation, it would be a good choice - we will discuss this)

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/CheckIn.cs:
```diff
@@ -0,0 +1,44 @@
+public class CheckIn
+{
+    private Passenger _passenger;
+    private BoardingPass boardingPass;
```
```suggestion
    private BoardingPass _boardingPass;
```

batonkarton replied:
I believe you don't need _boardingPass as a field because it can be created if the passenger passes all validations.
Field is used when you use this information in some logic inside your class in different places for example - _passenger is good choice for field inside CheckIn class.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passenger.cs:
```diff
@@ -0,0 +1,32 @@
+        Passport = passport;  
+    }
+
+    public bool PassportIsValid() => Passport?.PassportIsValid() ?? false;
```
Do you actually need this method? You have the same method in the Passport class and Passenger class has pa rop with the Passport and this information.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passport.cs:
```diff
@@ -0,0 +1,24 @@
+        PassportDate = passportDate;
+        PassportExpirationDate = passportExpirationDate;
+
+        if (PassportExpirationDate <= DateTime.Today)
```
```suggestion
        if (PassportExpirationDate < DateTime.Today)
```
I believe PassportExpirationDate = DateTime.Today should work for users - the last day.
And I'm not sure about the next CW ("Error: Passport has expired!") - should the user know about it? your program is still working after this message, so that's not the error - looks like it's a warning or something like that.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/BoardingPass.cs:
```diff
@@ -0,0 +1,17 @@
+
+    public BoardingPass(Passenger passenger)
+    {
+        Passenger = passenger;
```
Again Prop and initializing from constructor - that doesn't make sense because your prop is public - will talk about that.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/VisaCheck.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public static bool RequiresVisa(Countries.CountryPassengers country)
+    {
+        return !Enum.IsDefined(typeof(Countries.CountryPassengers), country);
```
Here you used Enum class but not in Program :)

batonkarton replied:
This method will always be false because user can't pass something wrong - you add strong validation inside Program.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/VisaCheck.cs:
```diff
@@ -0,0 +1,10 @@
+using AirportSimulation;
+
```
Formatting

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/VisaCheck.cs:
```diff
@@ -0,0 +1,10 @@
+
+public class VisaCheck
+{
+    public static bool RequiresVisa(Countries.CountryPassengers country)
```
```suggestion
    public static bool IsVisaRequired(Countries.CountryPassengers country)
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/SecurityCheck.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public bool ValidateSecurity(Passenger passenger)
+    {
+        
```
Formatting

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/SecurityCheck.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class SecurityCheck
+{
+    public bool ValidateSecurity(Passenger passenger)
```
```suggestion
    public bool IsSecurityCheckPassed(Passenger passenger)
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/CheckIn.cs:
```diff
@@ -0,0 +1,44 @@
+        boardingPass = new BoardingPass(_passenger);
+    }
+
+    public void CheckInByAirport()
```
```suggestion
    public void StartCheckIn()
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/PassportControl.cs:
```diff
@@ -0,0 +1,16 @@
+
+public class PassportControl
+{
+    public bool ValidatePassport(Passenger passenger)
```
```suggestion
    public bool IsPassportValidationPassed(Passenger passenger)
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/SecurityCheck.cs:
```diff
@@ -0,0 +1,21 @@
+        
+        Console.WriteLine($"A safety check is underway for the passenger {passenger.Name}...");
+       
+        bool hasProhibitedItems = false;  
```
Let's ask the user about ProhibitedItems and add validation steps here because now it's always false.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Flight.cs:
```diff
@@ -0,0 +1,65 @@
+namespace AirportSimulation;
+
+public class Flight
```
Do you use it?

AliakseiDziardziyashenka replied:
resolved

batonkarton reviewed (changes requested):
JFI Console has some Color properties - you can add red color for exceptions or yellow for warnings and green for success validations and messages - let's try to make it more user friendly.

AliakseiDziardziyashenka committed:
MT-25 added review fixes2 (+3 squashed commits)
Squashed commits:
[fc75b96] MT-25 added review fixes
[58383da] MT-25 Created Airport simulation (homework)
[4a0480b] MT-25 Created Airport simulation (homework)

AliakseiDziardziyashenka committed:
MT-25 Created Airport simulation (homework)

AliakseiDziardziyashenka committed:
MT-25 Created Airport simulation (homework)

batonkarton commented:
![image](https://github.com/user-attachments/assets/d44ffd75-ef69-475a-b46d-990c7961b083)
Please, write meaningful comments for commits.

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/SecurityCheck.cs:
```diff
@@ -0,0 +1,36 @@
+
+public class SecurityCheck
+{
+    public bool IsSecurityCheckPassed(Passenger passenger)
```
Please always check the IntelliSence suggestions - they are usually useful.
You can make this method static and don't have to create an instance of the class - remember about memory.
![image](https://github.com/user-attachments/assets/2c996ab0-7f62-4b4e-9917-08e14d113cab)

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,110 @@
+            {
+                break;
+            }
+            else
```
'Else' statement is not required here - also you have suggestions about it from IntelliSense. In real life, you will see the next code.

```suggestion
            if (!string.IsNullOrWhiteSpace(passportNumber))
            {
                break;
            }

            Console.WriteLine("Passport number cannot be empty. Please enter a valid passport number.");
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,110 @@
+            {
+                break;
+            }
+            else
```
'Else' statement is not required here - also you have suggestions about it from IntelliSense.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,110 @@
+            {
+                break;
+            }
+            else
```
'Else' statement is not required here - also you have suggestions about it from IntelliSense.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,110 @@
+            {
+                break;
+            }
+            else
```
'Else' statement is not required here - also you have suggestions about it from IntelliSense.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Program.cs:
```diff
@@ -0,0 +1,110 @@
+                selectedCountry = (Countries.CountryPassengers)countryChoice;
+                break;
+            }
+            else
```
'Else' statement is not required here - also you have suggestions about it from IntelliSense.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/CheckIn.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        Console.WriteLine("Check In By Airport");
+
+        var passportControl = new PassportControl();
```
Do you need this unused variable?

batonkarton replied:
At line 10 you are using static PassportControl.IsPassportValidationPassed()

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Countries.cs:
```diff
@@ -0,0 +1,24 @@
+namespace AirportSimulation;
+
+public class Countries
```
This class can be enum initially. 
In this case, you will be able to use just YourEnumClass in all places - not **Countries**._CountryPassengers_ -> just **CountryPassengers** - also about naming for Enum, usually it is in singular meaning, for example, Country (not **Countries**), PassportType (not **PassportTypes**)

I suggest changing Countries.cs -> PassengerCountry.cs (is it destination countries or where is the passenger from?)
```suggestion
namespace AirportSimulation;

public enum PassengerCountry
{
    Belarus,
    Russia,
    USA,
    Germany,
    France,
    Italy,
    Japan,
    China,
    Australia,
    UK,
    Canada,
    India,
    Brazil,
    SouthKorea,
    Spain,
    Mexico
}
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passport.cs:
```diff
@@ -0,0 +1,24 @@
+
+public class Passport
+{
+    public string? PassportNumber { get; set; }
```
Candidate for incapsulation - PassportNumber is used only in class Passport.
![image](https://github.com/user-attachments/assets/f11b656f-8979-4340-8417-5ae5f13e7d03)

batonkarton replied:
Do you need here nullable type(string?)? Are you expecting null?

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passport.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public string? PassportNumber { get; set; }
+    public DateTime PassportDate { get; set; }
+    public DateTime PassportExpirationDate { get; }
```
Candidate for incapsulation - PassportExpirationDate is used only in class Passport.
(take a look at suggestion)
![image](https://github.com/user-attachments/assets/1a5a5feb-6969-439f-b650-0a4d1d290a63)

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/BoardingPass.cs:
```diff
@@ -0,0 +1,10 @@
+namespace AirportSimulation;
+
+public class BoardingPass(Passenger passenger)
```
Do you need this unused code?

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/CheckIn.cs:
```diff
@@ -0,0 +1,26 @@
+        }
+
+        var securityCheck = new SecurityCheck();
+        if (!securityCheck.IsSecurityCheckPassed(passenger))
```
IsSecurityCheckPassed() could be static and you could reduce memory allocation and line 16(var securityCheck = new SecurityCheck();) will be removed.

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/CheckIn.cs:
```diff
@@ -0,0 +1,26 @@
+        if (!PassportControl.IsPassportValidationPassed(passenger, passport))
+        {
+            Console.WriteLine("\u274c The passenger did not pass the passport check.");
+            return;
```
MINOR: return; usually separated by a line. Not going to duplicate this comment for all code.
```suggestion
            Console.WriteLine("\u274c The passenger did not pass the passport check.");
            
            return;
```

AliakseiDziardziyashenka replied:
resolved

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/SecurityCheck.cs:
```diff
@@ -0,0 +1,36 @@
+        Console.ResetColor();
+        return true;
+    }
+}
```
MINOR: if you remember, I talked about the meaningful block of the code to a better understanding of the logic. So here are a lot of new lines without a reason that makes these code much more complicated than it is.

I would suggest refactoring and making logic blocks (comments are just for your understanding).
```suggestion
public static bool IsSecurityCheckPassed(Passenger passenger)
    {
        Console.WriteLine(
            $"A safety check is underway for the passenger {passenger.FirstName} {passenger.LastName}...");
        
        //block with validating answer
        Console.WriteLine("Do you have any prohibited items? (yes/no)");
        var input = Console.ReadLine()?.ToLower();
        var hasProhibitedItems = false;
        if (input == "yes")
        {
            hasProhibitedItems = true;
        }
        else if (input != "no")
        {
            Console.WriteLine("Invalid input. Please enter 'yes' or 'no'.");
            
            return false;
        }

        if (hasProhibitedItems)
        {
            Console.WriteLine($"\u274c {passenger.FirstName} {passenger.LastName}: Prohibited items found!");
            
            return false;
        }
        
        //block with result of security check
        Console.ForegroundColor = ConsoleColor.Green; 
        Console.WriteLine($"\u2705 {passenger.FirstName} {passenger.LastName}: Security check passed.");
        Console.ResetColor();
        
        return true;
    }
```

AliakseiDziardziyashenka replied:
resolved

AliakseiDziardziyashenka committed:
MT-25 added review fixes

AliakseiDziardziyashenka committed:
MT-25 added review fixes2

AliakseiDziardziyashenka committed:
Merge branch 'MT-24-Topic-9-HomeWork' of https://github.com/batonkarton/MentoAD into MT-24-Topic-9-HomeWork

* 'MT-24-Topic-9-HomeWork' of https://github.com/batonkarton/MentoAD:
  MT-25 added review fixes2
  MT-25 added review fixes
  MT-25 Created Airport simulation (homework)
  MT-25 Created Airport simulation (homework)

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Passenger.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public string FirstName { get; } = firstName ?? throw new ArgumentNullException(nameof(firstName));
+    public string LastName { get; } = lastName ?? throw new ArgumentNullException(nameof(lastName));
+    public DateTime DateOfBirth { get; } = dateOfBirth;
```
Property DateOfBirth is never used. It could be deleted.

batonkarton commented on MentoringHomeWork/HW9/AirportSimulation/Country.cs:
```diff
@@ -0,0 +1,21 @@
+namespace AirportSimulation;
+
+public enum CountryPassengers
```
Inconsistency with the name of the .cs file - **Country.cs** but `public enum CountryPassengers`

batonkarton reviewed (dismissed):
Approve with suggestions.

AliakseiDziardziyashenka committed:
MT-25 added review fixes 3
